pallets/treasury-buyout-extension/src/lib.rs

```diff
@@ -0,0 +1,525 @@
+		/// Emits `Buyout` event when successful.
+		#[pallet::call_index(0)]
+		#[pallet::weight((<T as pallet::Config>::WeightInfo::buyout(), Pays::No))]
+		pub fn buyout(
+			origin: OriginFor<T>,
+			asset: CurrencyIdOf<T>,
+			amount: Amount<BalanceOf<T>>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+			Self::do_buyout(who, asset, amount)?;
+			Ok(().into())
+		}
+
+		/// Allows root to update the buyout limit.
+		///
+		/// Parameters
+		///
+		/// - `origin`: Origin must be root.
+		/// - `limit`: New buyout limit. If None, then buyouts are not limited.
+		///
+		/// Emits `BuyoutLimitUpdated` event when successful.
+		#[pallet::call_index(1)]
+		#[pallet::weight(<T as pallet::Config>::WeightInfo::update_buyout_limit())]
+		pub fn update_buyout_limit(
+			origin: OriginFor<T>,
+			limit: Option<BalanceOf<T>>,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+			match limit {
+				Some(limit) => BuyoutLimit::<T>::put(limit),
+				None => BuyoutLimit::<T>::kill(),
+			}
+			Self::deposit_event(Event::<T>::BuyoutLimitUpdated { limit });
+			Ok(().into())
+		}
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Attempt to exchange native token to native token
+		WrongAssetToBuyout,
+		/// Buyout limit exceeded for the current period
+		BuyoutLimitExceeded,
+		/// One of transacted currencies is missing price information
+		NoPrice,
+		/// The treasury balance is too low for an operation
+		InsufficientTreasuryBalance,
+		/// The account balance is too low for an operation
+		InsufficientAccountBalance,
+		/// Exchange failed
+		ExchangeFailure,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Buyout event
+		Buyout {
+			who: AccountIdOf<T>,
+			buyout_amount: BalanceOf<T>,
+			asset: CurrencyIdOf<T>,
+			exchange_amount: BalanceOf<T>,
+		},
+		/// Buyout limit updated event
+		BuyoutLimitUpdated { limit: Option<BalanceOf<T>> },
+	}
+
+	/// Stores limit amount user could by for a period.
+	/// When `None` - buyouts are not limited
+	#[pallet::storage]
+	pub type BuyoutLimit<T: Config> = StorageValue<_, BalanceOf<T>, OptionQuery>;
+
+	/// Stores amount of buyouts (amount, block number of last buyout)
+	#[pallet::storage]
+	pub type Buyouts<T: Config> =
+		StorageMap<_, Blake2_128Concat, AccountIdOf<T>, (BalanceOf<T>, u32), ValueQuery>;
+}
+
+impl<T: Config> Pallet<T> {
+	/// Ensures that buyout limit is not exceeded for the current buyout period
+	fn ensure_buyout_limit_not_exceeded(
+		account_id: &AccountIdOf<T>,
+		buyout_amount: BalanceOf<T>,
+	) -> DispatchResult {
+		if let Some(buyout_limit) = BuyoutLimit::<T>::get() {
+			let buyout_period = T::BuyoutPeriod::get();
+			// Get current block number
+			let now = <frame_system::Pallet<T>>::block_number().saturated_into::<u32>();
+			let current_period = now
+				.checked_div(buyout_period)
+				.and_then(|n| Some(n.saturating_mul(buyout_period)))
+				.unwrap_or_default();
```

Comment:
What actually is the expected outcome here? Rounding? We are dividing and multiplying with the same value after all.

Reply:
I think it is because of rounding. First division if it has a remainder it will get truncated, multiplying by the number again will get you a smaller number. If one period has passed (lets say exactly), then it will not get truncated, which is then used int he condition bellow.
To me, it is the same as doing `now - buyout_period > last_buyout`, is it @bogdanS98 ?

Reply:
Yes, it is rounding. We use this to get the block when the current period started. For example if current block is `150000` and the buyout period is `7200` blocks then `current_period = (150000 / 7200) * 7200` which after the rounded division will become `current_period = 20 * 7200` therefore `current_period = 144000 `. Maybe I should refactor `current_period` and call it `current_period_start_block` to be clearer.

Reply:
@gianfra-t It is not actually the same as `now - buyout_period > last_buyout` since we're calculating which is the block number at which the last buyout period started and checking if the last buyout was executed at some block before that so that we can reset the buyout amount limit for that caller. If we were doing `now - buyout_period > last_buyout` then we would always assume that `now` is the end of a buyout period which is not always true.

Reply:
No I was not suggesting to change! Like you say it is similar but not exactly the same, and the solution implemented it makes more sense if we think in terms of periods.

Reply:
Nice, thanks for explaining.

I'm also okay to leave it like this. We could rename some variables to make it a little clearer, e.g. we can rename `now` -> `current_block_number` and `current_period` -> `current_period_start_number`, because the name 'period' to me indicates some range but here we only calculate the starting block of that range.